### DISCUSSION TITLE
Studying the Jacobian

### DISCUSSION THREAD

**Discussion by charlieowen3065** [idaholab/moose#27204] [Q&A General]
Hello.

For a project I am working on, I'd like to better understand the Jacobian in MOOSE, as well as how uncertainty propagates to the solution. 

I am having trouble amending it before going to PETSc though, and was wondering what a good method to do so was?

so far, I have added a function to "FEProblemSolve" that attempts to simply add a value to one of the elements in the sparse matrix. I also chose an element that was both non-zero and not part of the BC.

Below is the code I've ammended:
```
bool
FEProblemSolve::solve()
{
	// This loop is for nonlinear multigrids (developed by Alex)
	std::cout << "SOLVE : START" << std::endl;
	for (MooseIndex(_num_grid_steps) grid_step = 0; grid_step <= _num_grid_steps; ++grid_step)
	{
		std::cout << "SimpleAlter :: START" << std::endl;
		simple_jacobian_alter();
		std::cout << "SimpleAlter :: END" << std::endl;
		
		std::cout << "FEProblemSolve : (1)" << std::endl;
		_problem.solve(_nl.number());
		std::cout << "FEProblemSolve : (2)" << std::endl;

		if (_problem.shouldSolve())
		{
			if (_problem.converged(_nl.number()))
				_console << COLOR_GREEN << " Solve Converged!" << COLOR_DEFAULT << std::endl;
			else
			{
				_console << COLOR_RED << " Solve Did NOT Converge!" << COLOR_DEFAULT << std::endl;
				return false;
			}
		}
		else
			_console << COLOR_GREEN << " Solve Skipped!" << COLOR_DEFAULT << std::endl;

		if (grid_step != _num_grid_steps)
			_problem.uniformRefine();
	}
	std::cout << "SOLVE : START" << std::endl;
	
	return _problem.converged(_nl.number());
}


void 
FEProblemSolve::simple_jacobian_alter()
{
	// Petsc stuff
	auto & petsc_options = _problem.getPetscOptions();
	auto & pars = _problem.solverParams();
	Moose::PetscSupport::petscSetOptions(petsc_options, pars);
	//ExecFlagType flag = EXEC_INITIAL;
	ExecFlagType flag = _problem.getCurrentExecuteOnFlag();
	_problem.execute(flag);
	
	
	auto & jacobian = static_cast<ImplicitSystem &>(_nl.system()).get_system_matrix();
	// Assemble matrix
	_problem.computeJacobian(*_nl.currentSolution(), jacobian, 0);
	
	int i = 2;
	int j = 0;
	
	Real new_val = 10000;
	
	ExecFlagType flag_here = _problem.getCurrentExecuteOnFlag();
	std::cout << "Current Flag: " << flag_here << std::endl;
	std::cout << "JACOBIAN INFO: " << std::endl;
	
	jacobian.add(i, j, new_val);
	jacobian.print(std::cout, true);
}
```

And, when I run this, I get the error:

```
Time Step 0, time = 0

Time Step 1, time = 0.05, dt = 0.05
SAVING JACOBIAN
SOLVE : START
SimpleAlter :: START
Current Flag: NONE
JACOBIAN INFO:
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Object is in wrong state
[0]PETSC ERROR: Not for unassembled matrix
[0]PETSC ERROR: WARNING! There are unused option(s) set! Could be the program crashed before usage or a spelling mistake, etc!
[0]PETSC ERROR:   Option left: name:-i value: simple_heat_conduction_problem.i source: command line
[0]PETSC ERROR:   Option left: name:-ksp_converged_reason value: ::failed source: code
[0]PETSC ERROR:   Option left: name:-options_left value: 0 source: code
[0]PETSC ERROR:   Option left: name:-snes_converged_reason value: ::failed source: code
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.20.3, unknown
[0]PETSC ERROR: /home/nesquid/a/owen53/projects/malamute/malamute-opt on a  named nesquid.ecn.purdue.edu by owen53 Wed Mar 27 13:44:04 2024
[0]PETSC ERROR: Configure options --with-64-bit-indices --with-cxx-dialect=C++17 --with-debugging=no --with-fortran-bindings=0 --with-mpi=1 --with-openmp=1 --with-shared-libraries=1 --with-sowing=0 --download-fblaslapack=1 --download-hypre=1 --download-metis=1 --download-mumps=1 --download-ptscotch=1 --download-parmetis=1 --download-scalapack=1 --download-slepc=1 --download-strumpack=1 --download-superlu_dist=1 --with-hdf5-dir=${PREFIX} --with-make-np=16 --COPTFLAGS=-O3 --CXXOPTFLAGS=-O3 --FOPTFLAGS=-O3 --with-x=0 --with-ssl=0 --with-mpi-dir=/home/nesquid/a/owen53/miniforge/envs/moose AR=${PREFIX}/bin/x86_64-conda-linux-gnu-ar RANLIB=${PREFIX}/bin/x86_64-conda-linux-gnu-ranlib CFLAGS="-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include   -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include   -march=nocona -mtune=haswell" CXXFLAGS="-fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem ${PREFIX}/include  -march=nocona -mtune=haswell" CPPFLAGS="-DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include" FFLAGS="-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include   -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include   -I/home/nesquid/a/owen53/miniforge/envs/moose/include" FCFLAGS="-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include   -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/nesquid/a/owen53/miniforge/envs/moose/include   -I/home/nesquid/a/owen53/miniforge/envs/moose/include" LDFLAGS="-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/nesquid/a/owen53/miniforge/envs/moose/lib -Wl,-rpath-link,/home/nesquid/a/owen53/miniforge/envs/moose/lib -L/home/nesquid/a/owen53/miniforge/envs/moose/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/nesquid/a/owen53/miniforge/envs/moose/lib -Wl,-rpath-link,/home/nesquid/a/owen53/miniforge/envs/moose/lib -L/home/nesquid/a/owen53/miniforge/envs/moose/lib" --prefix=/home/nesquid/a/owen53/miniforge/envs/moose
[0]PETSC ERROR: #1 MatGetRow() at /data/civet0/build/conda/conda-bld/moose-petsc_1706652072497/work/src/mat/interface/matrix.c:566
libMesh terminating:
Not for unassembled matrix
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
Start: Wed Mar 27 13:44:03 EDT 2024
End: Wed Mar 27 13:44:05 EDT 2024

```

I have also tried inside the `Output` and `Predictor` type objects, but was seeing a similar error. I think it has something to do with the current executable flag being `NONE`, but maybe it is something else.

I know that the current school of thought is that small changes to the Jacobian will cause the solution to diverge, but I would like to test it out myself. Is there another place that I should try and change it?

I found the call to the libMesh::System object in "TimeIntegrator" (`_nl.system().solve();`), but don't see where I could properly access the Jacobian.

Ideally as well, I'd be able to access it at each non-linear iteration, but am just trying to start with the initial Jacobian.

Any advice?

Thanks!

Charlie

**Comment by GiudGiud** [comment]
Hello

Did you try modifying the kernels (especially the computeQpJacobian routine, only for non-AD kernels) to suit your needs first? The kernels could have enough information to perturb the Jacobian in the desired manner since they have enough information to compute the contribution to the Jacobian in the first place.

With regards to this specific error, you should pull up the part of the libmesh source code that is giving you this
```
libMesh terminating:
Not for unassembled matrix
```
error.

Altering the Jacobian before the solve might not work since we'll write to it and clear it a couple times during the solve. 

Guillaume

**Reply by charlieowen3065** [reply]
Hi Guillaume!

I have not tried modifying a kernel yet - I'll do that now!

Would using kernels allow me to modify it not only at the top of each time-step, but also at the iterations? I may be mistaken, but I assumed that the kernels were used to build the matrix using equations coming from the physics, then passed that information through the solver into PETSc, where it is then solved. Is there a back-and-forth between the kernels and PETSc? Or does the kernel only run once?

I will also look into libMesh for my specific error. 

Thanks!

Charlie

**Reply by GiudGiud** [reply]
> Is there a back-and-forth between the kernels and PETSc?

there is for PJFNK. PETSc needs to know how to compute a residual all the time
for Newton, there is a back and-forth on nonlinear iterations only iirc

**Reply by charlieowen3065** [reply]
Great - Thanks!

I'll try this out!

**Reply by lindsayad** [reply]
To be clear: there is a callback to the MOOSE residual evaluation for every linear iteration when running PJFNK. This does not happen when using NEWTON. Both NEWTON and PJFNK call for residual evaluations (potentially multiple times if using a line search) in the outer nonlinear iteration

**Reply by charlieowen3065** [reply]
Thank you for the responses.

I have been looking into the MOOSE Kernels, and I agree that this looks like where I should be working, but I am little confused on the mechanisms involved.

I see  that, in the kernel, there are functions to compute the 4x4 local Jacobian for a given element, and these functions use the Qp-computations given by specific kernel-types. However, I am lost on to where these local matrices are then used? 

I think most of my confusion revolves around this tagging system used. Does this system work by giving each element a unique tag, assigning these local matrices to the appropriate tag, then somewhere there is a method that understands how these tagged elements relate to one another in space and creates the full matrix using many tagged matrices? Then, in doing so, we can compute each local Jacobian for each kernel, add them up into one "tagged" local matrix? Meaning each local matrix will have the full physics involved when it moves to the full matrix?

Is this how that works? Also, if you could point me in the direction of where the local matrices are accumulated into one, I'd really appreciate it.

Thanks!

**Reply by GiudGiud** [reply]
> However, I am lost on to where these local matrices are then used?

these local matrices are contributed to the global Jacobian.

> Does this system work by giving each element a unique tag,

no. Tags are on per-matrix or per-vector basis. Once you create a tag for let's say a matrix A, you can give this tag to a few kernels to isolate their contributions to the Jacobian in matrix A. 

You can follow the chain of calculation by looking at computeQpJacobian, then computeJacobian, then one of the Jacobian loop (in loops/) onElement().

**Reply by charlieowen3065** [reply]
I see.

Is there a way to add my own loop into MOOSE? I try to refrain from touching source code where able. 

I was looking through the code, and I have not yet found where I could call my own personal loop in place of the, say, "computeJacobianThread" loop. Any suggestions on where to look?

Thanks!

**Reply by grmnptr** [reply]
Also, if you alter a matrix you can call `matrix.close()`. So between add and print. Also, `matrix.print()` should be enough.

**Reply by GiudGiud** [reply]
> Is there a way to add my own loop into MOOSE? I try to refrain from touching source code where able.

Yes feel free to create your own loop if you want. You will need to call it though. Use the residual loop as an example for creating the code for a new loop, and for executing that loop

**Reply by maxnezdyur** [reply]
@GiudGiud Correct me if I am wrong.
If this is just for learning and only run in serial, it may be easier for him to use GeneralUO and perform any actions to alter the Jacobian in `JacobianSetup`. We call `JacobianSetup` after we clear the Jacobian but before we add any elemental contributions. I can see this being easier for just an initial study problem but not good for production code.

**Reply by GiudGiud** [reply]
Yeah that’s a great idea.

**Reply by charlieowen3065** [reply]
Great! Thank you so much!

I'll start with the user object and see how much I can do there!

Thanks!

Charlie

**Reply by charlieowen3065** [reply]
Out of curiosity, would it be possible to set up a new loop using the user-object?

Maybe turning off the solver(?) and running a loop until convergence within my user object, updating solution, residual, and jacobian, and then allowing the problem to advance to the next timestep within the `execute()` function?

This may be a wonky way of doing this, but would this be possible/suggested as a preliminary test of adding a new loop into MOOSE? Just to get initial data on convergence and runtime?

Thanks!

**Reply by GiudGiud** [reply]
Hello

yes it s possible to nest an entire loop in a user object. A lot of the GeneralUOs do this. 

Turning off the solver and choosing when to advance to the next steps are typically not responsibilities of a user object. The Terminator UO does things similar to this but there are some complexities so I would not recommend that. 

Instead, have the UserObject provide some data to a postprocessor. The postprocessor makes these continuation / convergence decisions, and the executioner can take the convergence criterion from a postprocessor

Don’t worry too much about doing this in a wonky way. User objects are meant to allow you to do anything. 

Guillaume

**Reply by charlieowen3065** [reply]
For an initial test, I developed a full loop within the UserObject that, for a given timestep, takes in the previous solution as an initial guess, calculates the residual and jacobian using `_fe_problem.computeResidual(*solution, *residual, 0);` and `_fe_problem.computeJacobian(*solution, *jacobian, 0);`, then used `jacobain.lu_solve(resdiaul.scale(-1), deltaSolution);`, and then `solution.add(1, deltaSolution` to get my new solution vector.

to turn off the normal moose solver for the time-being, I used
```
[Problem]
	solve = false
[]
```

However, I am seeing that my solution blows up and my residual grows very quickly, and the system does not near convergence.

I have followed some of the flow, and right at the beginning, I am seeing 2 issues:

1) My Jacobian is not correct. It is not being updated properly, and I am not sure why. Do I need to add in my own "computeJacobian" loop? I am using the 'computeJacobian' function from FEProblemBase, and I also used the function from NonlinearSystemBase. For both, they are able to eventually figure a Jacobian close to what it should be, but by the time they do, the residual and solution have messed up and grown too large.

2) My residual is too high. For the first timestep, it should be ~0 everywhere, but I am seeing residuals of 1x and 2x my initial temperature conditions (this is a simple heat conduction problem for now).

Have I used a wrong function/used a function wrong? I am confused as to where I should look next.

I have posted my whole script below:

```
void
RomLoopUserObject_DEBUG::initialize()
{
	_jac_matrix = &static_cast<ImplicitSystem &>(_nl_sys.system()).get_system_matrix();
	_residual = &_nl_sys.RHS();
	//_curr_sol = &*_nl_sys.currentSolution();
	_curr_sol = &_nl_sys.solution();
}

void
RomLoopUserObject_DEBUG::execute()
{
	
	mainLoop();
}

void RomLoopUserObject_DEBUG::finalize() {}

void
RomLoopUserObject_DEBUG::mainLoop()
{
	std::cout << "LOOP START" << std::endl;

	bool is_converged = false;
	iter = 0;
	
	double tol = 1.0;
	int max_iter = 10;
	
	nNodes = _curr_sol->size();
	
	DenseVector<Number> S(nNodes);
	for (int i=0; i<nNodes; i++)
		S(i) = _curr_sol->el(i);
	
	NumericVector<Number> * R_numeric(_residual);
	for (int i=0; i<nNodes; i++)
		R_numeric->set(i, 0);

	while (!is_converged && iter < max_iter)
	{
		SparseMatrix<Number> * J(_jac_matrix);
		NumericVector<Number> * S_numeric = DenseToNumeric(S); 
		
		_fe_problem.computeJacobian(*S_numeric, *J, 0);
		_fe_problem.computeResidual(*S_numeric, *R_numeric, 0);
		
		DenseVector<Number> R = NumericToDense(*R_numeric);
		
		R.scale(-1.0);
		
		DenseMatrix<Number> J_dense = SparseToDenseMatrix(J);
		DenseVector<Number> delS;
		J_dense.lu_solve(R, delS);
		
		S.add(1, delS);
		
		updateSolution(S);
		
		NumericVector<Real> * S_updated_numeric = DenseToNumeric(S);
		
		Real res = computeResidual(S_updated_numeric, R_numeric);
		
		is_converged = res <= tol;
		
		std::cout << "Iteration " << iter << std::endl;
		std::cout << "    Residual: " << res << std::endl;
		if (is_converged)
			std::cout << "\n" << "SOLUTION CONVERGED (" << res << ")" << std::endl;
		
		iter++;	
	}
	
	if (!is_converged)
	{
		std::cout << "ERROR: DID NOT CONVERGE" << std::endl;
	}
}

NumericVector<Number> *
RomLoopUserObject_DEBUG::DenseToNumeric(DenseVector<Real> & D)
{
	int s = D.size();
	
	NumericVector<Number> * N(_residual);	
	
	N->init(0);
	N->init(s);
		
	for (int i=0; i<s; i++)
	{
		N->set(i, D.el(i));
		
	}
	
	return N;
}

DenseVector<Number>
RomLoopUserObject_DEBUG::NumericToDense(NumericVector<Real> & N)
{
	int s = N.size();
	DenseVector<Real> D(s);
	
	for (int i=0; i<s; i++)
	{
		D(i) = N.el(i);
	}
	
	return D;
}

DenseMatrix<Number>
RomLoopUserObject_DEBUG::SparseToDenseMatrix(SparseMatrix<Number> * S)
{
	int n = S->n();
	int m = S->m();
	
	std::vector<std::vector<Real>> D_mat;
	for (int i=0; i<n; i++)
	{
		std::vector<Real> row_i;
		for (int j=0; j<m; j++)
		{
			row_i.push_back(S[0](i,j));
		}
		D_mat.push_back(row_i);
	}
	
	DenseMatrix<Number> D(n, m);
	
	for (int i=0; i<n; i++)
		for (int j=0; j<m; j++)
			D(i, j) = D_mat[i][j];
	
	return D;
}

void
RomLoopUserObject_DEBUG::updateSolution(DenseVector<Real> new_solution)
{
	
	
	auto & local_soln = _nl_sys.solution();
	for (auto i=0; i<nNodes; i++)
	{
		local_soln.set(i, new_solution.el(i));
	}
	local_soln.close();
	_nl_sys.update();
	
	//NumericVector<Real> * S_rec_n = DenseToNumeric(new_solution);
	//_nl_sys.setSolution(*S_rec_n);
	//_nl_sys.update();
	
}

Real
RomLoopUserObject_DEBUG::computeResidual(NumericVector<Real> * S, NumericVector<Real> * R)
{
	_fe_problem.computeResidual(*S, *R, 0);
	auto err = R->l2_norm();
	return err;
}
```

and the residual output:

```
Time Step 0, time = 0

Time Step 1, time = 0.05, dt = 0.05
SAVING JACOBIAN
 Solve Skipped!
LOOP START
Iteration 0
    Residual: 1.70162e+09
Iteration 1
    Residual: 1.54117e+11
Iteration 2
    Residual: 5.68891e+14
Iteration 3
    Residual: 2.10026e+18
Iteration 4
    Residual: 7.75471e+21
Iteration 5
    Residual: 2.86351e+25
Iteration 6
    Residual: 1.05746e+29
Iteration 7
    Residual: 3.90536e+32
Iteration 8
    Residual: 1.44239e+36
Iteration 9
    Residual: 5.32755e+39
ERROR: DID NOT CONVERGE

Outlier Variable Residual Norms:
  temp: 5.327554e+39

```

**Reply by maxnezdyur** [reply]
What is the purpose of `S[0]` in the below?
When you look at D matrix is it empty?
```
DenseMatrix<Number>
RomLoopUserObject_DEBUG::SparseToDenseMatrix(SparseMatrix<Number> * S)
{
	int n = S->n();
	int m = S->m();
	
	std::vector<std::vector<Real>> D_mat;
	for (int i=0; i<n; i++)
	{
		std::vector<Real> row_i;
		for (int j=0; j<m; j++)
		{
			row_i.push_back(S[0](i,j));
		}
		D_mat.push_back(row_i);
	}
	
	DenseMatrix<Number> D(n, m);
	
	for (int i=0; i<n; i++)
		for (int j=0; j<m; j++)
			D(i, j) = D_mat[i][j];
	
	return D;
}
```
I think you wanted to do `(*S)(i,j)`

**Reply by charlieowen3065** [reply]
I got `S[0}(i, j)` more or less through trail-and-error working with the SparseMatrix object, but yes, that is a cleaner version of what I was trying to do.

However, the two methods ended up working the same - the vectors and matrices I receive from MOOSE are moving through my loop properly, and I have made sure to log and test most everything I can.

I am pretty sure that it is those two compute functions that are my source of error. Everything else seems to be working properly, but when I call those functions, I get unexpected results.

By "unexpected", I mean that I have run this same input file through the standard MOOSE, and I am comparing my results to those results. As my loop continues, the differences grow wider and wider, so right now I am just trying to figure out how to get from "Timestep begin" to the first iteration.

It is of note that the jacobian that is used for the first iteration is the same one I saved during the regular MOOSE's "Timestep Begin", and on my second iteration, the jacobian is similar to that of what it should be on the first iteration.

Also, for my residual, when it goes into the loop, it is a vector of zeros (correct), and for the first calculation, some areas remain zero, and some raise (which is expected), but there are also some near the borders which should remain zero, but instead increase to 300-600 (my T_infinity = 300K). However, but the second iteration, when the residual should be winding down, it, along with my solution vector, begins growing at every node.

I am still not sure what the issue is - my guess was that I was using one of the functions wrong, but I am not so sure. The initial residual is what I would expect it to be, and it takes the jacobian an extra iteration to get to near what it should be, but it does go in the right direction. But after the first iteration, everything goes haywire.

Do you notice anything else in my code that might be causing an issue? I am very confused, but will continue looking through the data and trying to find a solution.

Thanks!

Charlie

**Reply by GiudGiud** [reply]
if you have a MOOSE reference simulation to compare to, then I would encourage you to print the residual vector and the Jacobian.
You can pass `-ksp_view_pmat` to the petsc options to see it. 

And compare the MOOSE one to the one you are generating manually. 

At this point you are developing internals of MOOSE. The way we (and most likely PETSc folks did this at this level) do this is to have continuous integration and unit testing. 
For every routine you make you should create a few tests to ensure that it behaves as expected. This includes all the little conversion utilities you made. This will ensure they are correct now and remain correct as you keep debugging your problem
